File: Golang/first-class-citizen/first-class-citizen.go
package main

import "fmt"

func main() {
	// Kemampuan untuk memperlakukan fungsi sebagai nilai
	greet := func() {
		fmt.Println("Naveria Works")
	}

	greet()

	myKisah(waifu1)
	myKisah(waifu2)

	double := func(x int) int { return x * 2 }
	execute(double, 5)

	// // Metode 3 : Return dari fungsi lain
	// fungsiYangDipecah := get_nama()
	// fungsiYangDipecah()
}

// Kemampuan untuk meneruskan fungsi sebagai argumen
func waifu1() string {
	return "Chinatsu Kano"
}

func waifu2() string {
	return "Anzu Hanashiro"
}
// Di bahasa lain, "func() yang tidak mengembalikan nilai dan tak memiliki parameter disebut fungsi void"	
func myKisah(waifu func() string) {
	fmt.Println("Halo mY", waifu())
}

// Kemampuan untuk mengembalikan fungsi dari fungsi lain
// func get_nama() func() {
// 	return func() {
// 		fmt.Println("Halo King!")
// 	}
// }


// package main

// import "fmt"
	
// func main() {
// 	my_slice := []string{"Kano", "Anzu", "Naoko"}

// 	fmt.Println(my_slice, len(my_slice), cap(my_slice))

// 	new_slice := make([]int, 4, 10)

// 	fmt.Println(new_slice)
// 	new_slice[0] = 1

// 	fmt.Println(new_slice)

// 	// for i; i < 10; i ++ {
		
// 	// }
// }

// Fungsi yang menerima fungsi lain sebagai parameter
func execute(f func(int) int, val int) {
    fmt.Println("Result:", f(val))
}

File: Golang/first-class-citizen/first-class-citizen.js
// Kemampuan untuk memperlakukan fungsi sebagai nilai
let greet = function() {
    console.log("Naveria Works")
}

greet()

// Kemampuan untuk meneruskan fungsi sebagai argumen
function waifu1() {
    return "Chinatsu Kano"
}

function waifu2() {
    return "Anzu Hanashiro"
}

function myKisah(waifu) {
    console.log("Halo mY", waifu())
}

let istri_pertama = myKisah(waifu1)
let istri_kedua = myKisah(waifu2)

// Kemampuan untuk mengembalikan fungsi dari fungsi lain
let getNama = function() {
    return function() {
        console.log("Halo Kingg!")
    }
}

// Tanda kurung pertama merujuk ke fungsi anonim yang pertama, tepatnya ke fungsi itu sendiri.
// Tanda kurung kedua merujuk ke fungsi yang dikembalikan, yaitu fungsi anonim yang menyetak ke konsol
getNama()()

// Contoh Lain 1
let getNama1 = function() {
    return function() {
        console.log("Halo King!")
    }
}

// 1. Jalankan fungsi luar dan kemudian simpan hasilnya
let fungsiYangDikembalikan = getNama1()

// 2. Jalankan fungsi yang sudah disimpan
fungsiYangDikembalikan() 

// Contoh lain 2
function buatSapaan(salam) {
    let fungsiSapaan = function(nama) {
        console.log(`Halo, ${salam} Kak ${nama}!`)
    }
    
    return fungsiSapaan
}

let sapaan_pagi = buatSapaan("Selamat Pagi")
let sapaan_malam = buatSapaan("Selamat Malam")

sapaan_pagi("Kano")
sapaan_malam("Anzu")

// Panggil fungsi secara lansgung
buatSapaan("Woiii!")("Liaaa") // Pertama akan mengisi parameter salam dan dilanjutkan dengan parameter nama.